arroyo/processing/strategies/dead_letter_queue/policies/produce.py

```diff
@@ -0,0 +1,49 @@
+import json
+
+from arroyo.backends.kafka.consumer import KafkaPayload, KafkaProducer
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessages,
+)
+from arroyo.types import Topic
+from arroyo.utils.metrics import get_metrics
+
+
+class ProduceInvalidMessagePolicy(DeadLetterQueuePolicy):
+    """
+    Produces given InvalidMessages to a dead letter topic.
+
+    Meant to be used as a baseclass for policies needing to produce
+    invalid messages to a dead letter topic.
+    """
+
+    def __init__(self, producer: KafkaProducer, dead_letter_topic: Topic) -> None:
+        self.__metrics = get_metrics()
+        self.__dead_letter_topic = dead_letter_topic
+        self.__producer = producer
+
+    def handle_invalid_messages(self, e: InvalidMessages) -> None:
+        """
+        Produces a message to the given dead letter topic for each
+        invalid message in the form:
+        {
+            "topic": <original topic the bad message was produced to>,
+            "reason": <why the message(s) are bad>
+            "timestamp": <time at which exception was thrown>,
+            "message": <original bad message>
+        }
+        """
+        for message in e.messages:
+            data = json.dumps(
+                {
+                    "topic": e.topic,
+                    "reason": e.reason,
+                    "timestamp": e.timestamp,
+                    "message": message,
+                }
+            ).encode("utf-8")
+            payload = KafkaPayload(key=None, value=data, headers=[])
+            self.__producer.produce(
+                destination=self.__dead_letter_topic, payload=payload
+            )
```

Comment:
This would not work due to the asynchronous nature of the producer as you are never waiting for the callback.
Arroyo generally solves this problem by having separate `submit` `poll` `join` method. Where submit schedules more work to be done asynchronously. `poll` gets the callbacks that are being executed and `join` ensures that you never close the strategy before all callbacks are executed.
We need to do something asynchronous here as well.
So you have two options:
- call flush after this method or, when using the arroyo producer, wait on the future that you receive from the producer api. Please do not do that, while we do not expect too many messages here, waiting for the callback at every message is horrible for performance.
- Make the `DeadLetterQueue` policy more asynchronous and add a `join` (or you can call it differently if you want) method (in the parent class), just as we have in the processing strategies in arroyo. That method is meant to ensure that all the asynchronous work has completed before closing the strategy and exiting. When executing `handle_invalid_messages` you will call `produce` just like here. But you will store the response (which is a future). If somebody calls `join` on the `DeadLetterQueue` processing step, that one would call `join` on the DLQ policy before completing. So you are sure that we always wait for all callbacks. 

There will be an additional step. Like all asynchronous system based on a queue, every time you call `produce` you are accumulating work and keeping track of it but you continue immediately. This means that the calls to produce can come at a higher pace than the callbacks are invoked. So there has to be a limit on the size of your queue (the number of result object you keep track of). When you reach that limit and somebody calls `produce` you will have to wait for the queue to empty itself (the callbacks being invoked)

Reply:
Alright, I added a `join` method to the DLQ policies, along with a queue of up to 10 futures to the produce policy. I wired up the calls to the method similarly to how they work in ProcessingSteps, let me know if it looks good.

Reply:
> with a queue of up to 10 futures
I would suggest you set the queue size by default to 5000